PostgreSQL store implementation

123joshuawu opened the pull request.

gordom6 commented on app/kg/conf/application.conf:
```diff
@@ -5,6 +5,11 @@ kgStore = "neo4j"
 neo4j.password = "nC1aB4mji623s2Zs"
 neo4j.uri = "bolt://mcs-neo4j:7687"
 neo4j.user = "neo4j"
+slick.dbs.default.profile = "slick.jdbc.PostgresProfile$"
```
Please keep these alphabetized.

123joshuawu replied:
Fixed

gordom6 commented on build.sbt:
```diff
@@ -50,7 +50,10 @@ lazy val kgLib =
         "io.github.tetherless-world" %% "twxplore-test" % twxploreVersion % Test,
         "me.tongfei" % "progressbar" % "0.8.1",
-        "org.neo4j.driver" % "neo4j-java-driver" % "4.0.1"
+        "org.neo4j.driver" % "neo4j-java-driver" % "4.0.1",
```
And these.

123joshuawu replied:
Fixed

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -0,0 +1,91 @@
+  lazy val KgEdges = TableQuery[KgEdgeTable]
+  lazy val KgSources = TableQuery[KgSourceTable]
+
+  private class KgNodeTable(tag: Tag) extends Table[KgNode](tag, "KgNodes") {
```
Usual SQL convention is singular table and column names, with everything lower snake_case.

123joshuawu replied:
Ah okay changed

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -0,0 +1,91 @@
+  lazy val KgSources = TableQuery[KgSourceTable]
+
+  private class KgNodeTable(tag: Tag) extends Table[KgNode](tag, "KgNodes") {
+    def id = column[String]("ID", O.PrimaryKey)
```
I came up in the school that you should always have "id" be an autoincrement integer primary key and have separate UNIQUE constraints on "natural" keys like what we have in node.id. But others argue that it's fine to just have the "natural" key be the primary key if you have one. So you can leave this.

123joshuawu replied:
Thinking about it, that makes sense. I'm going to go ahead and change to integer primary keys to see how that pans out

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -0,0 +1,91 @@
+
+    type KgNodeTableColumnTypes = (String, Option[Char], Option[Int])
+
+    def * = (id, pos, wordNetSenseNumber) <> (create, extract.lift)
```
This Slick syntax is delightfully unintuitive.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -0,0 +1,91 @@
+  class KgEdgeTable(tag: Tag) extends Table[KgEdge](tag, "KgEdges") {
+    def id = column[String]("ID", O.PrimaryKey)
+//    labels: List[String],
+    def objectNodeId = column[String]("OBJECT_NODE_ID")
```
The s, p, o should form a composite unique constraint. Fine to do that later.

123joshuawu replied:
Cool, I will keep this in mind

123joshuawu replied:
Added a `unique_constraint` to `kg_edge` https://github.com/tetherless-world/mcs-apps/blob/0d12ee70978280c2a69a72b7a072d23fe8f79461/lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala#L30

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -0,0 +1,91 @@
+    def id = column[Int]("ID", O.PrimaryKey, O.AutoInc)
+
+    def kgNodeId = column[String]("KG_Node_ID")
+    def kgSourceId = column[String]("KG_SOURCE_ID")
```
Source ID should be the primary key here?

123joshuawu replied:
I made this for representing the many-to-many relationship between `KgSource` and `KgNode`, in which case I think the primary key could be the composite of `kg_node_id` and `kg_source_id`?

gordom6 replied:
Yes, you're right.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgQueryStore.scala:
```diff
@@ -0,0 +1,48 @@
+import scala.concurrent.duration._
+
+@Singleton
+final class PostgresKgQueryStore @Inject()(dbConfigProvider: DatabaseConfigProvider)(implicit executionContext: ExecutionContext) extends AbstractPostgresKgStore(dbConfigProvider) with KgQueryStore {
```
Not sure what your plan is here vis a vis the ABC, but it looks fine for now.

gordom6 reviewed (commented):
Looks like you're on the right track.

123joshuawu commented:
@gordom6 I'd like to get your feedback on the schema before moving on

Here is what the tables look like right now:
![Screen Shot 2020-10-26 at 2 44 34 AM](https://user-images.githubusercontent.com/9285017/97142561-35e87900-1737-11eb-8b4f-9791c37347e3.png)

gordom6 reviewed (commented):
pos should be char(1)
integers should be smallint
kg_node_label needs page_rank, so it will have to be a separate table with a join instead of just the join you have
  node label table: kg_node_label ( label: string primary key; page_rank: float; )
  node <-> node label join table: kg_node_kg_node_label ( label: string; kg_node_id: string )
probably need to represent node label<->node label edges separately from node<->node edges, rather than in a huge edge table
  if you want to go with the single edge table, then object_node_id should be object_id and subject_node_id subject_id  
we may want to consider using surrogate keys (int id autoincrement) later to avoid having the natural id strings as foreign keys, which will make the indices larger; this design is fine for now
sentences can be a multiple value, so violating normalization, but fine to leave it for now
there is a very small set of edge labels; this would be a good case for using surrogate integer id's in the join, but fine to leave it for now, too

Overall this looks good as a start. I suspect we'll end up using integer id's for performance reasons, but I'd like to see how this schema performs before optimizing.

123joshuawu commented:
- Fixed column types  
- Changed `kg_node_label` to represent node label
- Added `kg_node_label_edge` for node label <-> node label edges

Updated diagram:
https://dbdesigner.page.link/GQ3U7bmCH6nfy7oi9
![Screen Shot 2020-10-27 at 2 17 21 AM](https://user-images.githubusercontent.com/9285017/97264068-98ef1400-17fa-11eb-883a-5033d4441ce5.png)

gordom6 commented:
Use `text` instead of `varchar(n)`, there's no reason to limit the length of those fields. I assume `text` can still be a primary key.

page_rank should be `real` (aka `float4`), those don't need 8-byte precision.

It'd be nice to have some marker of the join tables, since the names can get awkward. I've used `_x_` in the past, as in `kg_node_x_kg_source`. (You currently call it `kg_edge_source`, which is inconsistent with the other join table names.)

You could probably dispense with the kg_ prefixes entirely. We need them in the code because it's shared with the benchmark app, but the neo4j store doesn't use kg_ prefixes in its naming.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -6,28 +6,31 @@ import slick.jdbc.PostgresProfile
     def predicate = column[String]("predicate")
     def sentences = column[String]("sentences")
-    def subjectNodeId = column[String]("subject_node_id")
+    def subjectKgNodeId = column[String]("subject_node_id")
```
You need some NOT NULL constraints, too? Not sure how Slick handles this.

123joshuawu replied:
I found this in the docs http://scala-slick.org/doc/3.0.0/schemas.html
> NotNull, Nullable
Explicitly mark the column as nullable or non-nullable when creating the DDL statements for the table. Nullability is otherwise determined from the type (Option or non-Option). There is usually no reason to specify these options.

gordom6 reviewed (approved):
Looks like you're on the right path. Some small refinements to the schema, then I'd like to see more implementation.

gordom6 reviewed (commented):
Marked "approve" instead of "comment" by accident.

123joshuawu commented:
I thought `VARCHAR(254)` was the default mapping for `String` but it turns out `TEXT` is the default for Postgres

gordom6 reviewed (commented):
Changes look good. Please continue with implementation of the API.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -3,10 +3,16 @@ package io.github.tetherlessworld.mcsapps.lib.kg.stores.postgres
+  protected val SentencesDelimChar = '|'
+  protected val SentencesDelimString = SentencesDelimChar.toString
+
   lazy val kgEdges = TableQuery[KgEdgeTable]
```
Remove the prefix from internal constructs like this. (Keep it on the outer class names like AbstractPostgresKgStore and PostgresKgQueryStore.)

123joshuawu replied:
Gotcha, prefixes removed

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -16,16 +22,28 @@ abstract class AbstractPostgresKgStore(protected val dbConfigProvider: DatabaseC
   lazy val kgSources = TableQuery[KgSourceTable]
 
-  private class KgEdgeTable(tag: Tag) extends Table[(String, String, String, String, String)](tag, "edge") {
+  protected def runTransaction[R](a: DBIOAction[R, NoStream, Nothing]): Future[R] = {
```
final

123joshuawu replied:
Added

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -55,15 +73,15 @@ abstract class AbstractPostgresKgStore(protected val dbConfigProvider: DatabaseC
   }
 
-  private class KgNodeTable(tag: Tag) extends Table[(String, Option[Short], Option[Short], Option[Float], Option[Char], Option[Short])](tag, "node") {
+  private class KgNodeTable(tag: Tag) extends Table[KgNodeRow](tag, "node") {
```
Much nicer.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    override def clear(): Unit = {
+//      TODO
```
TRUNCATE

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala:
```diff
@@ -0,0 +1,81 @@
+//      TODO
+    }
+
+    final private def generateEdgeInsert(edge: KgEdge) =
```
private methods don't need to be marked final, that's implied, but it's fine to leave this.

123joshuawu replied:
Ah didn't think that through, that makes sense

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    final private def generateEdgeInsert(edge: KgEdge) =
+      List(kgEdges += edge.toRow) ++
```
The mix of kg* prefixes and names without is jarring. That's why I suggested doing without. The Kg* prefixes are really just so other apps can reuse this code without duplicate class names (QueryStore, CommandStore, ... -> KgQueryStore, KgCommandStore).

123joshuawu replied:
Yeah that makes sense, removed

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala:
```diff
@@ -0,0 +1,81 @@
+    override def putNodes(nodes: Iterator[KgNode]): Unit =
+      runSyncTransaction(DBIO.sequence(nodes.flatMap(generateNodeInsert)))
+
+    override def putPaths(paths: Iterator[KgPath]): Unit = {
```
We can just nuke paths from the API. I'll do that. No one is using it right now. Initially it made sense to have the paths integrated in the neo4j store, because the path id's could be assigned to edge properties.

Just skip implementing this.

gordom6 replied:
I filed a separate issue (for me) to remove this from the API.

123joshuawu replied:
Sounds good

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala:
```diff
@@ -0,0 +1,81 @@
+    override def putEdges(edges: Iterator[KgEdge]) =
+      runSyncTransaction(DBIO.sequence(edges.flatMap(generateEdgeInsert)))
+
+    override def putKgtkEdgesWithNodes(edgesWithNodes: Iterator[KgtkEdgeWithNodes]): Unit =
```
Mark all these methods final.

123joshuawu replied:
Added `final`

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala:
```diff
@@ -0,0 +1,81 @@
+
+import scala.concurrent.ExecutionContext
+
+class PostgresKgCommandStore @Inject()(dbConfigProvider: DatabaseConfigProvider)(implicit executionContext: ExecutionContext) extends AbstractPostgresKgStore(dbConfigProvider) with KgCommandStore {
```
`@Singleton`

123joshuawu replied:
Whoops, added

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala:
```diff
@@ -0,0 +1,81 @@
+class PostgresKgCommandStore @Inject()(dbConfigProvider: DatabaseConfigProvider)(implicit executionContext: ExecutionContext) extends AbstractPostgresKgStore(dbConfigProvider) with KgCommandStore {
+  import profile.api._
+
+  private final class PostgresKgCommandStoreTransaction extends KgCommandStoreTransaction {
```
May be able to do 1 Postgres transaction : 1 KgCommandStoreTransaction, which would be ideal. neo4j couldn't do that because it buffered transactions in memory. Postgres doesn't do that.

123joshuawu replied:
Gotcha, I will try to do that

gordom6 reviewed (commented):
Looks good.

gordom6 commented on docker-compose.yml:
```diff
@@ -48,8 +48,24 @@ services:
+    environment:
+      POSTGRES_USER: "mcs"
+      POSTGRES_PASSWORD: "7EAdu7jJvZNxxrNZ"
+      POSTGRES_DB: "kg"
```
This is why you don't need the table name prefixes. neo4j only supports multiple databases in the enterprise versions.

123joshuawu replied:
Ah interesting

gordom6 commented on docker-compose.yml:
```diff
@@ -48,8 +48,24 @@ services:
+      - postgres-data:/var/lib/postgresql/data
+
 volumes:
   neo4j-data:
```
You'll want to merge master. I've modified this docker-compose.yml a few times to support the ISI installation.

123joshuawu replied:
Gotcha, I will hold off merging for now to avoid cluttering up the reviews

gordom6 commented on docker-compose.yml:
```diff
@@ -48,8 +48,24 @@ services:
+    networks:
+      - mcs
+    ports:
+      - 5432:5432
```
Only open to localhost (127.0.0.1:5432:5432).

123joshuawu replied:
Gotcha changed

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/KgStoresModule.scala:
```diff
@@ -23,14 +23,15 @@ final class KgStoresModule(environment: Environment, configuration: Configuratio
         bind(classOf[KgCommandStore]).to(classOf[TestKgStore])
         bind(classOf[KgQueryStore]).to(classOf[TestKgStore])
       }
-      case "postgres" => {
```
Alphabetized. I like it.

https://youtu.be/31M_MdSVxV8?t=102

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -11,147 +11,147 @@ abstract class AbstractPostgresKgStore(protected val dbConfigProvider: DatabaseC
-  protected def runTransaction[R](a: DBIOAction[R, NoStream, Nothing]): Future[R] = {
+  protected val SentencesDelimString: String = SentencesDelimChar.toString
+
+  protected lazy val edges = TableQuery[EdgeTable]
```
Looks much nicer.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -11,147 +11,147 @@ abstract class AbstractPostgresKgStore(protected val dbConfigProvider: DatabaseC
+  protected final case class NodeRow(id: String, inDegree: Option[Short], outDegree: Option[Short], pageRank: Option[Float], pos: Option[Char], wordNetSenseNumber: Option[Short])
 
-  private class KgEdgeTable(tag: Tag) extends Table[KgEdgeRow](tag, "edge") {
+  protected class EdgeTable(tag: Tag) extends Table[EdgeRow](tag, "edge") {
```
Can make these final, too, for good measure.

gordom6 replied:
My habit from Java is to make everything final and then remove it as needed.

123joshuawu replied:
That makes sense, I will keep that in mind

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala:
```diff
@@ -11,22 +11,23 @@ import play.api.db.slick.DatabaseConfigProvider
 
 import scala.concurrent.ExecutionContext
 
+@Singleton
```
Worth reading up on Guice if you haven't. It's a good example of Dependency Injection frameworks, which are a useful thing to have in your toolkit.

123joshuawu replied:
Yep am trying that out for the PostgresConfigProvider

123joshuawu commented:
@gordom6 am stuck on this error when I try to start the application, seems like my `@Inject` for the postgres config is not working but not sure why
```
CreationException: Unable to create injector, see the following errors:

1) Error injecting constructor, java.lang.NullPointerException
  at io.github.tetherlessworld.mcsapps.lib.kg.stores.postgres.PostgresKgCommandStore.<init>(PostgresKgCommandStore.scala:16)
  at io.github.tetherlessworld.mcsapps.lib.kg.stores.postgres.PostgresKgCommandStore.class(PostgresKgCommandStore.scala:16)
  while locating io.github.tetherlessworld.mcsapps.lib.kg.stores.postgres.PostgresKgCommandStore
  while locating io.github.tetherlessworld.mcsapps.lib.kg.stores.KgCommandStore
    for the 2nd parameter of io.github.tetherlessworld.mcsapps.lib.kg.stores.KgDataDirectoryLoader.<init>(KgDataDirectoryLoader.scala:24)
  at io.github.tetherlessworld.mcsapps.lib.kg.stores.KgStoresModule.configure(KgStoresModule.scala:44) (via modules: com.google.inject.util.Modules$OverrideModule -> stores.KgStoresModule -> io.github.tetherlessworld.mcsapps.lib.kg.stores.KgStoresModule)
  while locating io.github.tetherlessworld.mcsapps.lib.kg.stores.KgDataDirectoryLoader
```

gordom6 commented on lib/scala/kg/src/test/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgStoreSpec.scala:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  if (inTestingEnvironment) {
+    "The neo4j store" can {
```
Postgres.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -18,16 +19,34 @@ abstract class AbstractPostgresKgStore(protected val databaseConfigProvider: Pos
-  protected lazy val nodes = TableQuery[NodeTable]
-  protected lazy val nodeLabels = TableQuery[NodeLabelTable]
+  protected object nodes extends TableQuery(new NodeTable(_)) {
+    def withNodeLabels = for {
```
I think you probably need to pass the node id / node label you want into this. I'd have to see the SQL that results from using this, though, to see if it's pulling the whole table and then selecting what you want, or doing one operation.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/AbstractPostgresKgStore.scala:
```diff
@@ -18,16 +19,34 @@ abstract class AbstractPostgresKgStore(protected val databaseConfigProvider: Pos
+          .join(nodeLabels).on(_._2.nodeLabelLabel === _.label)
+    } yield (node, nodeLabel)
+
+    def withSources = for {
```
Same thing here, I'm concerned that this is going to select the whole join table first before filtering it.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgQueryStore.scala:
```diff
@@ -15,26 +15,67 @@ final class PostgresKgQueryStore @Inject()(configProvider: PostgresStoreConfigPr
+    runSyncTransaction((for {
+      ((node, source), (_, nodeLabel)) <- nodes
+        .withSources
+        .join(nodes.withNodeLabels)
```
I don't know Slick, but I suspect you want the results of the join combinator here rather the for expression.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgQueryStore.scala:
```diff
@@ -15,26 +15,67 @@ final class PostgresKgQueryStore @Inject()(configProvider: PostgresStoreConfigPr
+        .withSources
+        .join(nodes.withNodeLabels)
+        .on(_._1.id === _._1.id)
+        .filter(_._1._1.id === id)
```
LIMIT 1

gordom6 reviewed (commented):
Looks like it's on the right track. I'd like to see the generated SQL, though.

123joshuawu commented:
@gordom6 I think this wraps up the current pr. The command store should be complete, I will implement the remaining query store methods in the subsequent two prs in the following order:
1) PageRank 
- getNodeLabelContext()
2) Fulltext Search
- search()
- searchCount()
- searchFacets()

I'd like to get your review on the current state, then I will merge master into this branch to update the docker compose

gordom6 reviewed (commented):
Per Slack discussion, need some reworking on how transactions and operations are managed.

Also need to handle node source union semantics.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgQueryStore.scala:
```diff
@@ -1,29 +1,64 @@
+      .map(_.toString())
+  )
+
+  private implicit val getKgEdge = GetResult(r => KgEdge(
```
I'm not familiar with this syntax, but I'd guess it's tying you to a column order? which is not ideal. Should usually get columns by name and not by position.

123joshuawu replied:
Yeah it was tied to column order, changed to use column names instead

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgQueryStore.scala:
```diff
@@ -1,29 +1,64 @@
+  ))
+
+  private def toKgNodeLabels(rows: Seq[(NodeLabelRow, String, NodeRow, String, String)]) =
+    rows
```
Use of the tuple makes this hard to read, but not avoidable?

123joshuawu replied:
So I use it to convert the result rows into KgNodeLabels here:
https://github.com/tetherless-world/mcs-apps/blob/1853a98fba34e70168b627a7dbc6778ff42ad1e4/lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgQueryStore.scala#L93-L98

I think the query must return a tuple, so if I want to use a case class, it would have to be after the query is executed.

Maybe something like this?
```scala
case class KgNodeLabelQueryResultRow(nodeLabel: NodeLabelRow, nodeLabelSourceId: String...) {
  def apply(row: Tuple[NodeLabelRow, String, NodeRow, String, String]) =
    KgNodeLabelQueryResultRow(
      nodeLabel = row._1,
      ...
    )
}
...
toKgNodeLabels(runSyncTransaction(relatedNodeLabelWithNodeSourceAction).map(KgNodeLabelQueryResultRow(_))).toList
```

gordom6 replied:
Not worth the trouble in that case.

gordom6 commented on lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgQueryStore.scala:
```diff
@@ -35,42 +70,78 @@ final class PostgresKgQueryStore @Inject()(configProvider: PostgresStoreConfigPr
+      val relatedNodeLabels = toKgNodeLabels(runSyncTransaction(relatedNodeLabelWithNodeSourceAction)).toList
+
+      // TODO replace inner id order by with pageRank
+      val topEdgesQuery =
```
Once this is done please add some documentation as to what this is doing. I'd expect that the explanation of the SQL will take as many lines as the SQL itself.

123joshuawu replied:
Understood, will do

gordom6 reviewed (commented):
I assume the test failures are because of missing PageRank and missing fulltext search. If that's the case I'm ready to merge this PR so you can follow up with implementations of those things. Please confirm.

123joshuawu commented:
@gordom6 Yes, the test failures are due to the missing fulltext and pageRank
